cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/SourceControlManagementHttpHandler.java

```diff
@@ -68,23 +68,17 @@ public void setRepository(FullHttpRequest request, HttpResponder responder,
                             @PathParam("namespace-id") String namespaceId) throws Exception {
     checkSourceControlFeatureFlag();
     NamespaceId namespace = validateNamespaceId(namespaceId);
-    RepositoryConfigRequest repoRequest = getRepositoryConfigRequest(request);
-
+    
     try {
-      if (repoRequest == null || repoRequest.getConfig() == null) {
-        throw new RepositoryConfigValidationException("Repository configuration must be specified.");
-      }
-
-      repoRequest.getConfig().validate();
+      RepositoryConfigRequest repoRequest = validateAndGetRepoConfig(request);
       if (repoRequest.shouldTest()) {
-        // TODO: CDAP-20252, add the validate logic once the SourceControlManager module is ready
+        sourceControlService.validateRepository(namespace, repoRequest.getConfig());
+        responder.sendStatus(HttpResponseStatus.OK);
+        return;
       }
 
-      sourceControlService.setRepository(namespace, repoRequest.getConfig());
-      responder.sendJson(HttpResponseStatus.OK,
-                         GSON.toJson(
-                           new SetRepositoryResponse(
-                             String.format("Updated repository configuration for namespace '%s'.", namespaceId))));
+      RepositoryMeta repoMeta = sourceControlService.setRepository(namespace, repoRequest.getConfig());
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(repoMeta));
     } catch (RepositoryConfigValidationException e) {
```

Comment:
NIT:
Currently we are treating all errors as RepositoryConfigValidationException and returning 400. But for cases where we failed to get token from securestore due to network issues preferably be 500. Don't think we should be tackling this scenario right now.

Reply:
this is definitely an issue, though not one related to this PR. Can we open a jira so we don't forget about it?